File: Leetcode/easy/Code704.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int search(vector<int>& nums, int target) {
        std:cout << nums.size();
        int start = 0;
        int end = nums.size()-1;
        int mid = (start + end) / 2;
        while(start <=end){
            if (target<nums[start] || target>nums[end]) {
                return -1;
            }
            if (target == nums[start]) {
                return start;
            }
            if (target == nums[end]) {
                return end;
            }
            if (target == nums[mid]) {
                return mid;
            }
           
            if (target > nums[mid]  ) {
                start = mid+1;
            }
            if (target < nums[mid] )
            {
                end = mid-1;
            }
            mid = (start + end) / 2;
        }
        return -1;

    }
};
int main() {
    Solution s1;
    vector<int> t = { -1,0,3,5,9,12 };
    int s2=s1.search(t, 3);
    std:cout << s2;
}

File: Leetcode/easy/code704.js
/*
题目：给定一个 n 个元素有序的（升序）整型数组 nums 和一个目标值 target  ，写一个函数搜索 nums 中的 target，如果目标值存在返回下标，否则返回 -1

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/binary-search
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
*/
var search = function(nums, target) {
    let left = 0, right =  nums.length-1
    while(left <= right) {
        let mid = (left + right) >> 1
        if(nums[mid] === target) 
        return mid
        let isSmall = nums[mid] < target
        left = isSmall ? mid + 1 : 1
        right= isSmall ? r : mid - 1
    }   
    return -1
}